experiments/dutch_auctions/abaci.py

```diff
@@ -0,0 +1,69 @@
+""" Abaci Module
+    Class-based representation of the Abaci smart contract
+    (contains only what is necessary for the simulation)
+"""
+
+from pydss.pymaker.numeric import Wad, Rad
+from pydss.util import require
+
+
+class Abaci:
+    """
+    """
+
+    def __init__(self):
+        pass
+
+    def file(self, what, data):
+        pass
+
+    def price(self, top, dur):
+        """
+        top: initial price
+        dur: seconds since action has started
+        """
+        pass
+
+
+class LinearDecrease(Abaci):
+    """
+    tau = int
+    """
+
+    def __init__(self):
+        self.tau = 0
+        super().__init__()
+
+    def file(self, what, data):
+        if what == "tau":
+            self.tau = data
+        else:
+            raise Exception("LinearDecrease/file-unrecognized-param")
+
+    def price(self, top, dur):
+        if dur >= self.tau:
+            return 0
+        return top * (self.tau - dur) / self.tau  # TODO: need to cast to Rad
```

Comment:
Not sure what the correct way to port this to our implementation is, original code is abaci.sol line 85

Reply:
Ah yeah dealing w/ `Wad`/`Rad`/`Ray` is pretty annoying...
so for reference the OG is:
```solidity
rmul(top, mul(tau - dur, RAY) / tau)
```
to start, remember that `Ray`s are [fixed-point integers](https://en.wikipedia.org/wiki/Fixed-point_arithmetic) with 27 decimal places.
however, `tau` and `dur` are just normal integers, not fixed-point, as they represent some number of seconds. so the `mul(tau - dur, RAY)` call is there to just to convert `tau - dur` to a `Ray` (multiplies by `RAY = 10**27`). in our codebase you would do that with a `Ray.from_number((self.tau - dur) / self.tau)` (moving the ` / self.tau` inside `Ray.from_number` since it doesn't preserve order of arithmetic operations, i don't think)

at the same time, `rmul(x, y)` is pretty much given by `x * y / RAY`. this is for multiplying two `Ray`s, doing so will give you a fixed point integer w/54 decimal places, requiring you to divide by 10**27 to get a well-formatted `Ray`. Luckily for us, the `pymaker.numerics.Ray` class handles `Ray * Ray` multiplication on its own, and `top` is expected to be a `Ray` when passed in, so you can leave the final form as:
```python
return top * Ray.from_number((self.tau - dur) / self.tau)
```

a great example of similar wonkiness is our `pydss/cat.py:82`